R/write_output.R

```diff
@@ -0,0 +1,329 @@
+    obs_fact_table[["parameter"]],
+    levels = c("imputed_reports"),
+    labels = c("obs_reports")
+  ))
+
+  # Combine original fact_table with new 'obs_reports' rows
+  fact_table <- rbind(fact_table, obs_fact_table, fill = TRUE)
+  data.table::setnames(
+    fact_table,
+    old = c("parameter"),
+    new = c(".variable")
+  )
+
+  # Step 2: Extract desired parameters from the Stan object as posterior draws
+  stanfit <- fit[["estimates"]][["fit"]]
+  # Hacky workaround to avoid R CMD check NOTE on "no visible global binding"
+  # for variables in a dataframe evaluated via NSE. To use tidybayes, we need to
+  # use NSE, so giving these a global binding. The standard dplyr hacks
+  # (str2lang, .data prefix) didn't work here because it's not dplyr and we're
+  # not accessing a dataframe.
+  imputed_reports <- obs_reports <- R <- r <- time <- NULL # nolint
+  stan_draws <- tidybayes::gather_draws(
+    stanfit,
+    imputed_reports[time],
+    obs_reports[time],
+    R[time],
+    r[time],
+  ) |>
+    data.table::as.data.table()
+
+  return(list(stan_draws = stan_draws, fact_table = fact_table))
+}
+
+#' Post-process and merge posterior draws with a fact table.
+#'
+#' This function merges posterior draws with a fact table containing
+#' date-time-parameter combinations. It also standardizes parameter names and
+#' renames key columns.
+#'
+#' @param draws A data.table of posterior draws (either raw or summarized).
+#' @param fact_table A data.table of unique date-time-parameter combinations.
+#'
+#' @return A data.table with merged posterior draws and standardized parameter
+#' names.
+#' @noRd
+post_process_and_merge <- function(
+    draws,
+    fact_table,
+    geo_value,
+    model,
+    disease) {
+  # Step 1: Left join the date-time-parameter map onto the Stan draws
+  merged_dt <- merge(
+    draws,
+    fact_table,
+    by = c("time", ".variable"),
+    all.x = TRUE,
+    all.y = FALSE
+  )
+
+  # Step 2: Standardize parameter names
+  data.table::set(merged_dt, j = ".variable", value = factor(
+    merged_dt[[".variable"]],
+    levels = c("imputed_reports", "obs_reports", "R", "r"),
+    labels = c("latent_cases", "obs_cases", "Rt", "growth_rate")
+  ))
+
+  # Step 3: Rename columns as necessary
+  data.table::setnames(
+    merged_dt,
+    old = c(
+      ".draw", ".chain", ".variable", ".value", ".lower", ".upper", ".width",
+      ".point", ".interval", "date", ".iteration"
+    ),
+    new = c(
+      "_draw", "_chain", "_variable", "value", "_lower", "_upper", "_width",
+      "_point", "_interval", "reference_date", "_iteration"
+    ),
+    # If using summaries, skip draws-specific names
+    skip_absent = TRUE
+  )
+
+  # Metadata for downstream querying without path parsing or joins
+  data.table::set(merged_dt, j = "geo_value", value = factor(geo_value))
+  data.table::set(merged_dt, j = "model", value = factor(model))
+  data.table::set(merged_dt, j = "disease", value = factor(disease))
```

Comment:
When all of this gets saved to parquet, are these columns saved as categories? Will we run into issues where the dictionary encoding is different from one file to another like we did for the old NSSP gold data?

Reply:
Good q. Remind me -- did we issues reading with DuckDB or with polars?

Reply:
We saw issues when reading with anything other than duckdb, if we wanted to read more than one file at a time

Reply:
That said, I'm not sure how duckdb will write parquet files from R data.tables that have categorical columns. Maybe it will convert them to strings? 

Haven't been able to find anything in the duckdb docs yet about how it saves them.

Reply:
What's the rationale for the "_" prefix?

Reply:
The `.variable` prefix styling plays very poorly outside of R. It's particularly bad with DuckDB where you need to do a ton of escaping.

Reply:
I did a test locally, and duckdb seems to implicitly convert factor columns to UTF8 (string) columns

Reply:
I'm satisfied that this will be fine.